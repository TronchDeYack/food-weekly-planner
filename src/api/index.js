const recipeFake = {
  id: 'SPAGHETTI-ID',
  title: 'Spaghetti bolognaise',
  minutes: 20,
  caloriesLevel: 'medium',
  ingredients: [
    {
      id: 'PATE-ID',
      label: 'Pâtes',
      quantity: 200,
      unit: 'GR',
      type: 'STARCHY-FOOD',
    },
    {
      id: 'SAUCE-TOMATE-ID',
      label: 'Sauce Tomate',
      quantity: 200,
      unit: 'GR',
      type: 'CONDIMENT',
    },
    {
      id: 'STEAK-ID',
      label: 'Steak haché',
      quantity: 100,
      unit: 'GR',
      type: 'MEAT',
    },
    {
      id: 'OIL-ID',
      label: 'Huile d\'olives',
      quantity: 3,
      unit: 'CAC',
      type: 'CONDIMENT',
    },
  ],
};

const recipesFake = [
  recipeFake,
  {
    id: 'SOUPE-ID',
    title: 'Soupe à la tomate',
    minutes: 30,
    caloriesLevel: 'low',
    ingredients: [
      {
        id: 'TOMATE-ID',
        label: 'Tomates',
        quantity: 4,
        unit: 'UNIT',
        type: 'VEGETABLE',
      },
      {
        id: 'SALT-ID',
        label: 'Sel',
        quantity: 2,
        unit: 'CAC',
        type: 'CONDIMENT',
      },
    ],
  },
  {
    id: 'WELSH-ID',
    title: 'Welsh',
    minutes: 45,
    caloriesLevel: 'high',
    ingredients: [
      {
        id: 'CHEDDAR-ID',
        label: 'Cheddar',
        quantity: 400,
        unit: 'GR',
        type: 'DAIRY-PRODUCT',
      },
      {
        id: 'BEER-ID',
        label: 'Bière brune',
        quantity: 25,
        unit: 'CL',
        type: 'OTHER',
      },
      {
        id: 'HAM-ID',
        label: 'Jambon',
        quantity: 100,
        unit: 'GR',
        type: 'MEAT',
      },
      {
        id: 'BREAD-ID',
        label: 'Pain',
        quantity: 4,
        unit: 'SLICE',
        type: 'STARCHY-FOOD',
      },
    ],
  },
];

export function getRecipe(id) {
  console.log(`get recipe ${id}`);
  return Promise.resolve(recipeFake);
}

export function getRecipes() {
  return Promise.resolve(recipesFake);
}

export function getRecipesOfTheWeek() {
  return Promise.resolve([recipesFake[1], recipesFake[2]]);
}

export function getShoppingList() {
  return Promise.resolve([
    {
      id: 'BREAD-ID',
      label: 'Pain',
      quantity: 1,
      unit: 'unit',
      type: 'STARCHY-FOOD',
      bought: false,
    },
    {
      id: 'CHEDDAR-ID',
      label: 'Cheddar',
      quantity: 400,
      unit: 'GR',
      type: 'DAIRY-PRODUCT',
      bought: true,
    },
    {
      id: 'MILK-ID',
      label: 'Lait',
      quantity: 1,
      unit: 'L',
      type: 'DAIRY-PRODUCT',
      bought: false,
    },
    {
      id: 'BEER-ID',
      label: 'Bière brune',
      quantity: 25,
      unit: 'CL',
      type: 'OTHER',
      bought: false,
    },
    {
      id: 'HAM-ID',
      label: 'Jambon',
      quantity: 100,
      unit: 'GR',
      type: 'MEAT',
      bought: true,
    },
    {
      id: 'CHICKEN-ID',
      label: 'Piou piou',
      quantity: 200,
      unit: 'GR',
      type: 'MEAT',
      bought: false,
    },
  ]);
}

export function getIngredients() {
  return Promise.resolve([
    {
      id: 'PATE-ID',
      label: 'Pâtes',
      quantity: 200,
      unit: 'GR',
      type: 'STARCHY-FOOD',
    },
    {
      id: 'SAUCE-TOMATE-ID',
      label: 'Sauce Tomate',
      quantity: 200,
      unit: 'GR',
      type: 'CONDIMENT',
    },
    {
      id: 'STEAK-ID',
      label: 'Steak haché',
      quantity: 100,
      unit: 'GR',
      type: 'MEAT',
    },
    {
      id: 'OIL-ID',
      label: 'Huile d\'olives',
      quantity: 3,
      unit: 'CAC',
      type: 'CONDIMENT',
    },
    {
      id: 'TOMATE-ID',
      label: 'Tomates',
      quantity: 4,
      unit: 'UNIT',
      type: 'VEGETABLE',
    },
    {
      id: 'SALT-ID',
      label: 'Sel',
      quantity: 2,
      unit: 'CAC',
      type: 'CONDIMENT',
    },
    {
      id: 'CHEDDAR-ID',
      label: 'Cheddar',
      quantity: 400,
      unit: 'GR',
      type: 'DAIRY-PRODUCT',
    },
    {
      id: 'BEER-ID',
      label: 'Bière brune',
      quantity: 25,
      unit: 'CL',
      type: 'OTHER',
    },
    {
      id: 'HAM-ID',
      label: 'Jambon',
      quantity: 100,
      unit: 'GR',
      type: 'MEAT',
    },
    {
      id: 'BREAD-ID',
      label: 'Pain',
      quantity: 4,
      unit: 'SLICE',
      type: 'STARCHY-FOOD',
    },
  ]);
}
