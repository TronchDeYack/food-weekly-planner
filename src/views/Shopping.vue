<template>
  <v-expansion-panels>
    <v-expansion-panel v-for="shoppingCategory in shoppingCategories" :key="shoppingCategory.type">
      <v-expansion-panel-header>
        {{ getCategoryHeader(shoppingCategory) }}
      </v-expansion-panel-header>
      <v-expansion-panel-content>
        <v-list>
          <v-list-item-group multiple active-class="">
            <v-list-item
              v-for="ingredient in shoppingCategory.ingredients" :key="ingredient.id"
            >
              <template>
                <v-list-item-content>
                  <v-list-item-title>{{ ingredient.label }}</v-list-item-title>
                </v-list-item-content>
                <v-list-item-action>
                  <v-checkbox color="primary"
                    @click.native="toggleIngredient(ingredient.id)"
                    :input-value="ingredient.bought"
                    v-model="ingredient.bought"
                  />
                </v-list-item-action>
              </template>
            </v-list-item>
          </v-list-item-group>
        </v-list>
      </v-expansion-panel-content>
    </v-expansion-panel>
  </v-expansion-panels>
</template>

<style scoped>
.v-list-item--link:before {
  background-color: white !important;
}
</style>

<script>
import { getShoppingList } from '@/api';

export default {
  name: 'shopping',
  data() {
    return {
      ingredients: [],
    };
  },
  mounted() {
    getShoppingList().then((ingredients) => {
      this.ingredients = ingredients;
    });
  },
  methods: {
    getCategoryHeader(category) {
      return `${category.type} (${category.nbrBought}/${category.nbrToBuy})`;
    },
    toggleIngredient(ingredientId) {
      const ingredientIndex = this.ingredients.findIndex(
        ingredient => ingredient.id === ingredientId,
      );
      this.ingredients[ingredientIndex].bought = !this.ingredients[ingredientIndex].bought;
    },
  },
  computed: {
    shoppingCategories() {
      const categories = this.ingredients.reduce((accumulator, currentIngredient) => {
        const currentIngredientType = currentIngredient.type;
        const ingredientCategoryIndex = accumulator.findIndex(
          ingredient => ingredient.type === currentIngredientType,
        );
        const ingredientLabel = currentIngredient.unit === 'unit'
          ? `${currentIngredient.label} (${currentIngredient.quantity})`
          : `${currentIngredient.label} (${currentIngredient.quantity} ${currentIngredient.unit})`;

        if (ingredientCategoryIndex !== -1) {
          accumulator[ingredientCategoryIndex].ingredients.push({
            id: currentIngredient.id,
            label: ingredientLabel,
            bought: currentIngredient.bought,
          });
        } else {
          accumulator.push({
            type: currentIngredientType,
            ingredients: [{
              id: currentIngredient.id,
              label: ingredientLabel,
              bought: currentIngredient.bought,
            }],
          });
        }
        return accumulator;
      }, []);
      return categories.map(category => ({
        ...category,
        nbrToBuy: category.ingredients.length,
        nbrBought: category.ingredients.filter(ingredient => ingredient.bought).length,
      }));
    },
  },
};
</script>
