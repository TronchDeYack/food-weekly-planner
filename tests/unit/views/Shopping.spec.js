import { shallowMount } from '@vue/test-utils';

import Shopping from '@/views/Shopping.vue';
import { getShoppingList } from '@/api';

jest.mock('@/api');

describe('Shopping', () => {
  let wrapper;

  const ingredientsToShopping = [
    {
      id: 'BREAD-ID',
      label: 'Pain',
      quantity: 1,
      unit: 'unit',
      type: 'STARCHY-FOOD',
      bought: false,
    },
    {
      id: 'CHEDDAR-ID',
      label: 'Cheddar',
      quantity: 400,
      unit: 'GR',
      type: 'DAIRY-PRODUCT',
      bought: true,
    },
    {
      id: 'MILK-ID',
      label: 'Lait',
      quantity: 1,
      unit: 'L',
      type: 'DAIRY-PRODUCT',
      bought: false,
    },
    {
      id: 'BEER-ID',
      label: 'Bière brune',
      quantity: 25,
      unit: 'CL',
      type: 'OTHER',
      bought: false,
    },
    {
      id: 'HAM-ID',
      label: 'Jambon',
      quantity: 100,
      unit: 'GR',
      type: 'MEAT',
      bought: true,
    },
    {
      id: 'CHICKEN-ID',
      label: 'Piou piou',
      quantity: 200,
      unit: 'GR',
      type: 'MEAT',
      bought: true,
    },
  ];

  beforeEach(() => {
    getShoppingList.mockResolvedValue(ingredientsToShopping);
    wrapper = shallowMount(Shopping);
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('state', () => {
    it('Should initialize "ingredients" state to an empty array.', () => {
      wrapper = shallowMount(Shopping);
      expect(wrapper.vm.$data.ingredients).toEqual([]);
    });
  });

  describe('mounted', () => {
    it('Should get ingredients data from API and store them in "ingredients" state.', () => {
      expect(wrapper.vm.$data.ingredients).toEqual(ingredientsToShopping);
    });
  });

  describe('computed', () => {
    it('Should organize the ingredients by type.', () => {
      const expectedOrganizedIngredients = [
        {
          type: 'STARCHY-FOOD',
          ingredients: [
            {
              id: 'BREAD-ID',
              label: 'Pain (1)',
              bought: false,
            },
          ],
          nbrBought: 0,
          nbrToBuy: 1,
        },
        {
          type: 'DAIRY-PRODUCT',
          ingredients: [
            {
              id: 'CHEDDAR-ID',
              label: 'Cheddar (400 GR)',
              bought: true,
            },
            {
              id: 'MILK-ID',
              label: 'Lait (1 L)',
              bought: false,
            },
          ],
          nbrBought: 1,
          nbrToBuy: 2,
        },
        {
          type: 'OTHER',
          ingredients: [
            {
              id: 'BEER-ID',
              label: 'Bière brune (25 CL)',
              bought: false,
            },
          ],
          nbrBought: 0,
          nbrToBuy: 1,
        },
        {
          type: 'MEAT',
          ingredients: [
            {
              id: 'HAM-ID',
              label: 'Jambon (100 GR)',
              bought: true,
            },
            {
              id: 'CHICKEN-ID',
              label: 'Piou piou (200 GR)',
              bought: true,
            },
          ],
          nbrBought: 2,
          nbrToBuy: 2,
        },
      ];

      expect(wrapper.vm.shoppingCategories).toEqual(expectedOrganizedIngredients);
    });
  });

  describe('methods', () => {
    describe('getCategoryHeader', () => {
      it('Should return the type with the remaining ingredients number to buy and the total ingredients number.', () => {
        const category = {
          type: 'DAIRY-PRODUCT',
          ingredients: [
            {
              id: 'CHEDDAR-ID',
              label: 'Cheddar (400 GR)',
              bought: true,
            },
            {
              id: 'MILK-ID',
              label: 'Lait (1 L)',
              bought: false,
            },
          ],
          nbrBought: 1,
          nbrToBuy: 2,
        };

        const expectedHeader = 'DAIRY-PRODUCT (1/2)';

        expect(wrapper.vm.getCategoryHeader(category)).toEqual(expectedHeader);
      });
    });

    describe('toggleIngredient', () => {
      it('Should update "bought" property to true when checking the ingredient.', () => {
        wrapper.vm.toggleIngredient(ingredientsToShopping[2].id);

        expect(wrapper.vm.$data.ingredients[2].bought).toBe(true);
      });

      it('Should update "bought" property to false when unchecking the ingredient.', () => {
        wrapper.vm.toggleIngredient(ingredientsToShopping[1].id);

        expect(wrapper.vm.$data.ingredients[1].bought).toBe(false);
      });
    });
  });
});
