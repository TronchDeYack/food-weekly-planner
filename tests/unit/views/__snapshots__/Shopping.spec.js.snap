// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopping Should render. 1`] = `
<v-expansion-panels-stub activeclass="v-item--active">
  <v-expansion-panel-stub>
    <v-expansion-panel-header-stub expandicon="$expand">
      STARCHY-FOOD (0/1)
    </v-expansion-panel-header-stub>
    <v-expansion-panel-content-stub>
      <v-list-stub tag="div" tile="true">
        <v-list-item-group-stub activeclass="" multiple="true">
          <v-list-item-stub activeclass="" tag="div">
            <v-list-item-content-stub>
              <v-list-item-title-stub>Pain (1)</v-list-item-title-stub>
            </v-list-item-content-stub>
            <v-list-item-action-stub>
              <v-checkbox-stub color="primary" errorcount="1" errormessages="" messages="" rules="" successmessages="" backgroundcolor="" ripple="true" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" indeterminateicon="$checkboxIndeterminate" officon="$checkboxOff" onicon="$checkboxOn"></v-checkbox-stub>
            </v-list-item-action-stub>
          </v-list-item-stub>
        </v-list-item-group-stub>
      </v-list-stub>
    </v-expansion-panel-content-stub>
  </v-expansion-panel-stub>
  <v-expansion-panel-stub>
    <v-expansion-panel-header-stub expandicon="$expand">
      DAIRY-PRODUCT (1/2)
    </v-expansion-panel-header-stub>
    <v-expansion-panel-content-stub>
      <v-list-stub tag="div" tile="true">
        <v-list-item-group-stub activeclass="" multiple="true">
          <v-list-item-stub activeclass="" tag="div">
            <v-list-item-content-stub>
              <v-list-item-title-stub>Cheddar (400 GR)</v-list-item-title-stub>
            </v-list-item-content-stub>
            <v-list-item-action-stub>
              <v-checkbox-stub color="primary" errorcount="1" errormessages="" messages="" rules="" successmessages="" value="true" backgroundcolor="" ripple="true" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" inputvalue="true" indeterminateicon="$checkboxIndeterminate" officon="$checkboxOff" onicon="$checkboxOn"></v-checkbox-stub>
            </v-list-item-action-stub>
          </v-list-item-stub>
          <v-list-item-stub activeclass="" tag="div">
            <v-list-item-content-stub>
              <v-list-item-title-stub>Lait (1 L)</v-list-item-title-stub>
            </v-list-item-content-stub>
            <v-list-item-action-stub>
              <v-checkbox-stub color="primary" errorcount="1" errormessages="" messages="" rules="" successmessages="" backgroundcolor="" ripple="true" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" indeterminateicon="$checkboxIndeterminate" officon="$checkboxOff" onicon="$checkboxOn"></v-checkbox-stub>
            </v-list-item-action-stub>
          </v-list-item-stub>
        </v-list-item-group-stub>
      </v-list-stub>
    </v-expansion-panel-content-stub>
  </v-expansion-panel-stub>
  <v-expansion-panel-stub>
    <v-expansion-panel-header-stub expandicon="$expand">
      OTHER (0/1)
    </v-expansion-panel-header-stub>
    <v-expansion-panel-content-stub>
      <v-list-stub tag="div" tile="true">
        <v-list-item-group-stub activeclass="" multiple="true">
          <v-list-item-stub activeclass="" tag="div">
            <v-list-item-content-stub>
              <v-list-item-title-stub>Bi√®re brune (25 CL)</v-list-item-title-stub>
            </v-list-item-content-stub>
            <v-list-item-action-stub>
              <v-checkbox-stub color="primary" errorcount="1" errormessages="" messages="" rules="" successmessages="" backgroundcolor="" ripple="true" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" indeterminateicon="$checkboxIndeterminate" officon="$checkboxOff" onicon="$checkboxOn"></v-checkbox-stub>
            </v-list-item-action-stub>
          </v-list-item-stub>
        </v-list-item-group-stub>
      </v-list-stub>
    </v-expansion-panel-content-stub>
  </v-expansion-panel-stub>
  <v-expansion-panel-stub>
    <v-expansion-panel-header-stub expandicon="$expand">
      MEAT (2/2)
    </v-expansion-panel-header-stub>
    <v-expansion-panel-content-stub>
      <v-list-stub tag="div" tile="true">
        <v-list-item-group-stub activeclass="" multiple="true">
          <v-list-item-stub activeclass="" tag="div">
            <v-list-item-content-stub>
              <v-list-item-title-stub>Jambon (100 GR)</v-list-item-title-stub>
            </v-list-item-content-stub>
            <v-list-item-action-stub>
              <v-checkbox-stub color="primary" errorcount="1" errormessages="" messages="" rules="" successmessages="" value="true" backgroundcolor="" ripple="true" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" inputvalue="true" indeterminateicon="$checkboxIndeterminate" officon="$checkboxOff" onicon="$checkboxOn"></v-checkbox-stub>
            </v-list-item-action-stub>
          </v-list-item-stub>
          <v-list-item-stub activeclass="" tag="div">
            <v-list-item-content-stub>
              <v-list-item-title-stub>Piou piou (200 GR)</v-list-item-title-stub>
            </v-list-item-content-stub>
            <v-list-item-action-stub>
              <v-checkbox-stub color="primary" errorcount="1" errormessages="" messages="" rules="" successmessages="" value="true" backgroundcolor="" ripple="true" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" inputvalue="true" indeterminateicon="$checkboxIndeterminate" officon="$checkboxOff" onicon="$checkboxOn"></v-checkbox-stub>
            </v-list-item-action-stub>
          </v-list-item-stub>
        </v-list-item-group-stub>
      </v-list-stub>
    </v-expansion-panel-content-stub>
  </v-expansion-panel-stub>
</v-expansion-panels-stub>
`;
